service/collector/app.go

```diff
@@ -146,19 +153,59 @@ func (c *App) collectAppStatus(ctx context.Context, ch chan<- prometheus.Metric)
 	return nil
 }
 
-func (c *App) getTeam(ctx context.Context, app v1alpha1.App) (string, error) {
-	name := key.AppCatalogEntryName(app.Spec.Catalog, app.Name, app.Spec.Version)
+func (a *App) getOwningTeam(ctx context.Context, app v1alpha1.App, ownersYAML string) (string, error) {
+	owners := []owner{}
+
+	err := yaml.Unmarshal([]byte(ownersYAML), &owners)
+	if err != nil {
+		// If the YAML in the owners annotation is invalid log the error and
+		// fall back to trying the team annotation.
+		a.logger.Errorf(ctx, err, "could not parse owners YAML for app %q", key.AppName(app))
+		return "", nil
+	}
+
+	for _, o := range owners {
+		if key.CatalogName(app) == o.Catalog && a.provider == o.Provider {
+			return o.Team, nil
+		} else if key.CatalogName(app) == o.Catalog && o.Provider == "" {
+			return o.Team, nil
+		} else if o.Catalog == "" && a.provider == o.Provider {
+			return o.Team, nil
+		}
+	}
+
+	// If no owning team is found we fall back to the team annotation.
+	return "", nil
+}
 
-	ace, err := c.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogEntries(metav1.NamespaceDefault).Get(ctx, name, metav1.GetOptions{})
+func (a *App) getTeam(ctx context.Context, app v1alpha1.App) (string, error) {
+	var team string
+
+	name := key.AppCatalogEntryName(key.CatalogName(app), key.AppName(app), key.Version(app))
+
+	ace, err := a.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogEntries(metav1.NamespaceDefault).Get(ctx, name, metav1.GetOptions{})
 	if apierrors.IsNotFound(err) {
-		return c.defaultTeam, nil
+		return a.defaultTeam, nil
 	} else if err != nil {
 		return "", microerror.Mask(err)
 	}
 
-	team := key.AppCatalogEntryTeam(*ace)
+	// Owners annotation takes precedence if it exists.
+	ownersYAML := key.AppCatalogEntryOwners(*ace)
+
+	if ownersYAML != "" {
+		team, err = a.getOwningTeam(ctx, app, ownersYAML)
```

Comment:
Can we parse `yaml` first and give a parsed struct to `getOwingTeam`? Since we are using this pattern in app-operator (We parsed metadata and give it to the function as a parameter) I think that makes more sense.

Reply:
Yes that's cleaner. 👍

Reply:
This is done. PTAL